### PR TITLE
CODENVY-398; Make it possible to perform integration with customers ldap

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
 -  Adds possibility to migrate data from 3-rd party LDAP server;
 -  Adds possibility to authenticate users via 3-rd party LDAP server;

### What issues does this PR fix or reference?
#CODENVY-398

### Tests written?
Yes

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -854,6 +859,11 @@
                 <version>${org.flyway.version}</version>
             </dependency>
             <dependency>
+                <groupId>org.ldaptive</groupId>
+                <artifactId>ldaptive</artifactId>
+                <version>1.2.0</version>
```
Comment (sleshchenko) [inline pom.xml]: move version to properties

Comment (skabashnyuk) [inline pom.xml]: move version to property

File: wsmaster/codenvy-hosted-api-admin/src/main/java/com/codenvy/api/user/server/AdminUserCreator.java [hunk 2]
```diff
@@ -68,26 +78,53 @@
...
-            try {
-                adminUser =  userManager.create(new UserImpl(name, email, name, password, emptyList()), false);
-                LOG.info("Admin user '" + name + "' successfully created");
-            } catch (ConflictException cfEx) {
-                LOG.warn("Admin user creation failed", cfEx.getLocalizedMessage());
-                return;
+            if (shouldCreateAdmin) {
+                try {
+                    User adminUser = userManager.create(new UserImpl(name, email, name, password, emptyList()), false);
+                    grantSystemPermissions(adminUser.getId());
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-admin/src/main/java/com/codenvy/api/user/server/AdminUserCreator.java]: Will be user granted twice? Here and on event?

File: wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AuthenticationHandler.java [hunk 3]
```diff
@@ -24,13 +23,13 @@
  */
 public interface AuthenticationHandler {
     /**
-     * Authenticate user.
+     *  Check user password.
+     *
+     * @throws AuthenticationException - in case if login name and password are not matched.
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AuthenticationHandler.java]: please format this doc

File: wsmaster/codenvy-ldap/src/main/java/com/codenvy/ldap/sync/LdapSynchronizerService.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+@Path("/ldap/sync")
+@Api("/ldap/sync")
+public class LdapSynchronizerService {
+
+    @Inject
+    private LdapSynchronizer synchronizer;
+
+    @POST
+    @ApiOperation("Forces immediate users/profiles synchronization")
+    @ApiResponses({@ApiResponse(code = 200, message = "Synchronization successfully started"),
```
Comment (sleshchenko) [inline wsmaster/codenvy-ldap/src/main/java/com/codenvy/ldap/sync/LdapSynchronizerService.java]: Mb 204 would be better here

File: wsmaster/codenvy-ldap/src/test/java/com/codenvy/ldap/sync/LookupSelectorTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public class LookupSelectorTest {
+
+    private EmbeddedLdapServer server;
+    private ConnectionFactory  connFactory;
+
+    @BeforeClass
+    public void setUpServer() throws Exception {
+        (server = EmbeddedLdapServer.newDefaultServer()).start();
+
```
Comment (sleshchenko) [inline wsmaster/codenvy-ldap/src/test/java/com/codenvy/ldap/sync/LookupSelectorTest.java]: As for me too many empty lines here

File: wsmaster/codenvy-ldap/src/test/resources/META-INF/persistence.xml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            <property name="javax.persistence.jdbc.url" value="jdbc:h2:mem:"/>
+            <property name="javax.persistence.jdbc.user" value=""/>
+            <property name="javax.persistence.jdbc.password" value=""/>
+
+            <property name="eclipselink.exception-handler" value="org.eclipse.che.api.core.h2.jdbc.jpa.eclipselink.H2ExceptionHandler"/>
+            <property name="eclipselink.target-server" value="None"/>
+            <property name="eclipselink.ddl-generation" value="create-tables"/>
+            <property name="eclipselink.ddl-generation.output-mode" value="database"/>
+            <property name="eclipselink.logging.logger" value="DefaultLogger"/>
+            <property name="eclipselink.logging.level" value="FINE"/>
```
Comment (sleshchenko) [inline wsmaster/codenvy-ldap/src/test/resources/META-INF/persistence.xml]: Should we use SEVERE logging level here?

File: wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AuthenticationHandler.java (hunk 3 above)
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AuthenticationHandler.java]: OK

File: pom.xml (hunk 1 above)
Comment (mshaposhnik) [inline pom.xml]: Done

File: assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiServletModule.java [hunk 7]
```diff
@@ -96,6 +96,8 @@ protected void configureServlets() {
                "/preferences/*",
                "/license",
                "/license/*",
+               "/ldap/sync",
+               "/ldap/sync/*,
```
Comment (mshaposhnik) [inline assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiServletModule.java]: Yep - merge bug

File: wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AccessTicket.java [hunk 8]
```diff
@@ -171,7 +138,7 @@ public int hashCode() {
     public String toString() {
         final StringBuilder sb = new StringBuilder("AccessTicket{");
         sb.append("registeredClients=").append(registeredClients);
-        sb.append(", principal=").append(principal);
+        sb.append(", userId=").append(userId);
```
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/dao/authentication/AccessTicket.java]: Yup

File: wsmaster/codenvy-hosted-sso-server/src/test/java/com/codenvy/auth/sso/server/ticket/InMemoryTicketManagerTest.java [hunk 9]
```diff
@@ -30,7 +27,7 @@
 
 public class InMemoryTicketManagerTest {
     private final static String TOKEN = "123456789";
-    Subject subjectPrincipal = new SubjectImpl("user", "123", "t1", false);
+    String subjectPrincipal = "sdfsd345345";
```
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-sso-server/src/test/java/com/codenvy/auth/sso/server/ticket/InMemoryTicketManagerTest.java]: Much better

File: wsmaster/codenvy-ldap/src/test/java/com/codenvy/ldap/sync/LookupSelectorTest.java (hunk 5 above)
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/src/test/java/com/codenvy/ldap/sync/LookupSelectorTest.java]: Add few more:)

File: wsmaster/codenvy-ldap/src/test/resources/META-INF/persistence.xml (hunk 6 above)
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/src/test/resources/META-INF/persistence.xml]: Yes we can

File: wsmaster/codenvy-ldap/src/main/java/com/codenvy/ldap/sync/LdapSynchronizerService.java (hunk 4 above)
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/src/main/java/com/codenvy/ldap/sync/LdapSynchronizerService.java]: Ok

File: wsmaster/codenvy-ldap/pom.xml [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    is strictly forbidden unless prior written permission is obtained
+    from Codenvy S.A..
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <artifactId>codenvy-master-parent</artifactId>
+        <groupId>com.codenvy.onpremises.wsmaster</groupId>
+        <version>5.0.0-M4-SNAPSHOT</version>
```
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/pom.xml]: +

File: wsmaster/codenvy-ldap/IM.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+1) install IM CLI, execute: `bash <(curl -L -s https://get.codenvy.com) --im-cli`
+
+2) execute: `. ~/.bashrc`
+
+3) Download codenvy onprem bundle for tiaa.
+
+Just in case here is instruction how to build codenvy bundle:
+
+- clone following repos, checkout to branch `ldap`, and build them with `mvn clean install -DskipTests -Dskip-validate-sources`
```
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/IM.md]: No

File: wsmaster/codenvy-ldap/IM.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
+How to install codenvy onprem from custom bundle.
+
+1) install IM CLI, execute: `bash <(curl -L -s https://get.codenvy.com) --im-cli`
+
+2) execute: `. ~/.bashrc`
+
+3) Download codenvy onprem bundle for tiaa.
```
Comment (mshaposhnik) [inline wsmaster/codenvy-ldap/IM.md]: Nope

**Pull request merged**